internal/config/schedule.go

```diff
@@ -1,172 +1,156 @@
-		}
-	}
+func (r *RuntimeConfig) applyPendingSchedules() {
+	// Set of schedules (by id) which Rotation was altered and where RefreshRotations must be called.
+	updatedScheduleIds := make(map[int64]struct{})
+
+	incrementalApplyPending(
+		r,
+		&r.Schedules, &r.configChange.Schedules,
+		nil,
+		func(curElement, update *recipient.Schedule) error {
+			curElement.Name = update.Name
+			return nil
+		},
+		nil)
+
+	incrementalApplyPending(
+		r,
+		&r.scheduleRotations, &r.configChange.scheduleRotations,
+		func(newElement *recipient.Rotation) error {
+			schedule, ok := r.Schedules[newElement.ScheduleID]
+			if !ok {
+				return fmt.Errorf("rotation refers to unknown schedule %d", newElement.ScheduleID)
+			}
 
-	var rotationMemberPtr *recipient.RotationMember
-	stmt = r.db.BuildSelectStmt(rotationMemberPtr, rotationMemberPtr)
-	r.logger.Debugf("Executing query %q", stmt)
+			schedule.Rotations = append(schedule.Rotations, newElement)
+			updatedScheduleIds[schedule.ID] = struct{}{}
+			return nil
+		},
+		func(curElement, update *recipient.Rotation) error {
+			curElement.ActualHandoff = update.ActualHandoff
+			curElement.Priority = update.Priority
+			curElement.Name = update.Name
+			return nil
+		},
+		func(delElement *recipient.Rotation) error {
+			schedule, ok := r.Schedules[delElement.ScheduleID]
+			if !ok {
+				return fmt.Errorf("rotation refers to unknown schedule %d", delElement.ScheduleID)
+			}
 
-	var members []*recipient.RotationMember
-	if err := tx.SelectContext(ctx, &members, stmt); err != nil {
-		r.logger.Errorln(err)
-		return err
-	}
+			schedule.Rotations = slices.DeleteFunc(schedule.Rotations, func(rotation *recipient.Rotation) bool {
+				return rotation.ID == delElement.ID
+			})
+			updatedScheduleIds[schedule.ID] = struct{}{}
+			return nil
+		})
+
+	incrementalApplyPending(
+		r,
+		&r.scheduleRotationMembers, &r.configChange.scheduleRotationMembers,
+		func(newElement *recipient.RotationMember) error {
+			newElement.TimePeriodEntries = make(map[int64]*timeperiod.Entry)
+
+			var schedule *recipient.Schedule
+			for id, tmpSchedule := range r.Schedules {
+				ok := slices.ContainsFunc(tmpSchedule.Rotations, func(rotation *recipient.Rotation) bool {
+					return rotation.ID == newElement.RotationID
+				})
+				if ok {
+					newElement.ScheduleID = id
+					schedule = tmpSchedule
+					break
+				}
+			}
+			if schedule == nil {
+				return fmt.Errorf("schedule rotation member cannot be mapped to a schedule")
+			}
 
-	rotationMembersById := make(map[int64]*recipient.RotationMember)
-	for _, member := range members {
-		memberLogger := r.logger.With(zap.Object("rotation_member", member))
+			rotationId := slices.IndexFunc(schedule.Rotations, func(rotation *recipient.Rotation) bool {
+				return rotation.ID == newElement.RotationID
+			})
+			if rotationId < 0 {
+				return fmt.Errorf("schedule rotation member refers unknown rotation %d", newElement.RotationID)
+			}
```

Comment:
How can you end up in this situation? Isn't that schedule retrieved based on `newElement.RotationID` in the above loop and handled accordingly if it is nil? I'm also wondering why you added the `ScheduleID` field to `RotationMember` type.

Reply:
This is a case of "this should never happen", but I prefer a helpful error instead of a panic resulting in the next line.

Reply:
> but I prefer a helpful error instead of a panic resulting in the next line.

As far as I can see, there's no way that could result in a panic here, since you literally searched the `Schedule` via the rotation ID. This IMHO just adds some dead codes that shouldn't be needed at all.

Reply:
Sorry, I failed to get your point earlier as I was just looking at the quoted code section. However, when scrolling out to see the bigger picture, I get it.